cluster/images/etcd/migrate/migrate_client.go

```diff
@@ -0,0 +1,217 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"context"
+
+	clientv2 "github.com/coreos/etcd/client"
+	"github.com/coreos/etcd/clientv3"
+	"github.com/golang/glog"
+)
+
+// CombinedEtcdClient provides an implementation of EtcdMigrateClient using a combination of the etcd v2 client, v3 client
+// and etcdctl commands called via the shell.
+type CombinedEtcdClient struct {
+	cfg *EtcdMigrateCfg
+}
+
+// NewEtcdMigrateClient creates a new EtcdMigrateClient from the given EtcdMigrateCfg.
+func NewEtcdMigrateClient(cfg *EtcdMigrateCfg) (EtcdMigrateClient, error) {
+	return &CombinedEtcdClient{cfg}, nil
+}
+
+// Close closes the client and releases any resources it holds.
+func (e *CombinedEtcdClient) Close() error {
+	return nil
+}
+
+// SetEtcdVersionKeyValue writes the given version to the etcd 'etcd_version' key.
+// If no error is returned, the write was successful, indicating the etcd server is available
+// and able to perform consensus writes.
+func (e *CombinedEtcdClient) SetEtcdVersionKeyValue(version *EtcdVersion) error {
+	return e.Put(version, "etcd_version", version.String())
+}
+
+// Put write a single key value pair to etcd.
+func (e *CombinedEtcdClient) Put(version *EtcdVersion, key, value string) error {
+	if version.major == 2 {
+		v2client, err := e.clientV2()
+		if err != nil {
+			return err
+		}
+		_, err = v2client.Set(context.Background(), key, value, nil)
+		return err
+	}
+	v3client, err := e.clientV3()
+	if err != nil {
+		return err
+	}
+	defer v3client.Close()
+	_, err = v3client.KV.Put(context.Background(), key, value)
+	return err
+}
+
+// Get reads a single value for a given key.
+func (e *CombinedEtcdClient) Get(version *EtcdVersion, key string) (string, error) {
+	if version.major == 2 {
+		v2client, err := e.clientV2()
+		if err != nil {
+			return "", err
+		}
+		resp, err := v2client.Get(context.Background(), key, nil)
+		if err != nil {
+			return "", err
+		}
+		return resp.Node.Value, nil
+	}
+	v3client, err := e.clientV3()
+	if err != nil {
+		return "", err
+	}
+	defer v3client.Close()
+	resp, err := v3client.KV.Get(context.Background(), key)
+	if err != nil {
+		return "", err
+	}
+	kvs := resp.Kvs
+	if len(kvs) == 0 {
+		return "", fmt.Errorf("key not found: %s", key)
+	}
+	return string(kvs[0].Value), nil
```

Comment:
if length > 1 how do you know your index 0 is the right key/value pair? Wouldn't it be better to require an exact match until we understand how to grab the correct entry in the list?

Reply:
I'll replace this check with:
```
if len(kvs) != 1 {                                                                                                                                                                                                                                        
        return "", fmt.Errorf("expected exactly one value for key %s but got ", key, len(kvs))                                                                                                                                                                
     }
```

The API is only suppose to return 0 or 1 key for a Get request like this (since it's not a range or prefix get), but you're right, we should still check for it and document what we expect.